database/sqlite3.ksy

```diff
@@ -32,213 +32,431 @@ doc: |
     seq:
-      - id: len_payload
+      - id: p
         type: vlq_base128_be
       - id: row_id
         type: vlq_base128_be
       - id: payload
-        size: len_payload.value
-        type: cell_payload
-      # TODO: overflow
-  cell_table_interior:
+        type:
+          switch-on: '(p.value > _root.header.table_x ? 1 : 0)'
+          cases:
+            0: record
+            1: overflow_record(p.value, _root.header.table_x)
+  table_interior_cell:
     doc-ref: 'https://www.sqlite.org/fileformat.html#b_tree_pages'
     seq:
       - id: left_child_page
-        type: u4
+        type: btree_page_pointer
       - id: row_id
         type: vlq_base128_be
-  cell_index_leaf:
+  index_leaf_cell:
     doc-ref: 'https://www.sqlite.org/fileformat.html#b_tree_pages'
     seq:
-      - id: len_payload
+      - id: p
         type: vlq_base128_be
       - id: payload
-        size: len_payload.value
-        type: cell_payload
-      # TODO: overflow
-  cell_index_interior:
+        type:
+          switch-on: '(p.value > _root.header.index_x ? 1 : 0)'
+          cases:
+            0: record
+            1: overflow_record(p.value, _root.header.index_x)
+  index_interior_cell:
     doc-ref: 'https://www.sqlite.org/fileformat.html#b_tree_pages'
     seq:
       - id: left_child_page
-        type: u4
-      - id: len_payload
+        type: btree_page_pointer
+      - id: p
         type: vlq_base128_be
       - id: payload
-        size: len_payload.value
-        type: cell_payload
-  cell_payload:
+        type:
+          switch-on: '(p.value > _root.header.index_x ? 1 : 0)'
+          cases:
+            0: record
+            1: overflow_record(p.value, _root.header.index_x)
+  record:
     doc-ref: 'https://sqlite.org/fileformat2.html#record_format'
     seq:
-      - id: len_header_and_len
+      - id: header_length
         type: vlq_base128_be
-      - id: column_serials
-        size: len_header_and_len.value - 1
-        type: serials
-      - id: column_contents
+      - id: header
+        type: record_header
+        size: header_length.value - 1
+      - id: values
+        type: value(header.value_types[_index])
         repeat: expr
-        repeat-expr: column_serials.entries.size
-        type: column_content(column_serials.entries[_index])
-  serials:
+        repeat-expr: header.value_types.size
+  record_header:
     seq:
-      - id: entries
-        type: vlq_base128_be
+      - id: value_types
+        type: serial_type
         repeat: eos
-  serial:
+  serial_type:
+    -webide-representation: "{value_type:dec}"
     seq:
       - id: code
         type: vlq_base128_be
     instances:
-      is_blob:
-        value: 'code.value >= 12 and (code.value % 2 == 0)'
-      is_string:
-        value: 'code.value >= 13 and (code.value % 2 == 1)'
-      len_content:
+      value_type:
+        value: 'code.value >= 12 ? ((code.value % 2 == 0) ? 12 : 13 + _root.header.text_encoding - 1) : code.value'
```

Comment:
rename `code` to `raw_value`?
rename `value_type` to `type`?

raw_value = type + size

```
>>> value_type.code.value
23
>>> value_type.value_type
13
```

```
>>> serial_type.code.value
23
>>> serial_type.value_type
13
```

Reply:
can we have enum for type?

```
0: null
# big-endian twos-complement integers
1: int8
2: int16
3: int24
4: int32
5: int48
6: int64
7: float64 # big-endian IEEE 754-2008 64-bit float
8: int0 # the integer 0
9: int1 # the integer 1
# Reserved for internal use
10: internal10
11: internal11
12: blob
13: string
```

Reply:
Done

Reply:
> I bumped into a Kaitai limit here where I couldn't source the string encoding from an expression.

can you elaborate?

```
# line 465
  serial:
    # ...
    # Value is a string in the text encoding and (N-13)/2 bytes in length. The nul terminator is
    # not stored.
    13: string_utf8
    14: string_utf16_le
    15: string_utf16_be
```


at least add comment for 14 and 15

ideally avoid this workaround

Reply:
I had put the string value of the encoding (`UTF-8`, `UTF16-LE`, or `UTF-16BE`) in an instance attribute in the header. There didn't seem to be a way to reference that from the `encoding:` attribute of an `str` typed value. As a consequence I needed to have three distinct string types and a way to reference them. That's what those three enum constants are for.

Reply:
Thinking about this a bit more, you could get rid of these `string_...` enum values, replace `text_encoding` value in the header with another enum, and then switch on that in a single string type. I think you'll still need the three string types to actually be able to specify the encoding for Kaitai.

The resulting tree is going to be a bit uglier though. I was kind of optimising things for display in the KSV tree view and cross-referencing with the SQLite technical documentation.